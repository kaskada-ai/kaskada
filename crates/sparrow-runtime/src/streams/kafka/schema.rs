use schema_registry_converter::{
    async_impl::schema_registry::{get_schema_by_subject, SrSettings},
    schema_registry_common::SubjectNameStrategy,
};

use error_stack::{IntoReport, Result, ResultExt};

/// TODO: Pending Implementation
#[allow(unused)]
#[derive(Debug, derive_more::Display)]
pub enum Error {
    #[allow(dead_code)]
    AvroNotEnabled,
    AvroSchemaConversion,
    SchemaRequest,
    UnsupportedSchema,
}
impl error_stack::Context for Error {}

#[allow(unused)]
/// TODO: Pending implementation
pub async fn get_kafka_schema() -> Result<String, Error> {
    let sr_settings = SrSettings::new(String::from("http://localhost:8085"));
    let subject_name_strategy =
        SubjectNameStrategy::TopicNameStrategy(String::from("my-topic"), false);
    let schema = get_schema_by_subject(&sr_settings, &subject_name_strategy)
        .await
        .into_report()
        .change_context(Error::SchemaRequest)?;
    Ok(schema.schema)
}
