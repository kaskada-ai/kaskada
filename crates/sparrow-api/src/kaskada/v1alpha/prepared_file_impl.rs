use chrono::NaiveDateTime;

use crate::kaskada::v1alpha::PreparedFile;

#[derive(derive_more::Display, Debug)]
pub enum Error {
    #[display(fmt = "missing field '{field}")]
    MissingField { field: &'static str },
    #[display(fmt = "invalid timestamp {timestamp:?}")]
    InvalidTimestamp {
        timestamp: prost_wkt_types::Timestamp,
    },
}

impl error_stack::Context for Error {}

impl From<&PreparedFile> for PreparedFile {
    fn from(file: &PreparedFile) -> Self {
        file.clone()
    }
}

impl PreparedFile {
    pub fn min_event_time(&self) -> error_stack::Result<NaiveDateTime, Error> {
        let Some(timestamp) = self.min_event_time.as_ref() else {
            error_stack::bail!(Error::MissingField {
                field: "min_event_time"
            })
        };
        NaiveDateTime::from_timestamp_opt(timestamp.seconds, timestamp.nanos as u32).ok_or_else(
            || {
                error_stack::report!(Error::InvalidTimestamp {
                    timestamp: timestamp.clone()
                })
            },
        )
    }

    pub fn max_event_time(&self) -> error_stack::Result<NaiveDateTime, Error> {
        let Some(timestamp) = self.max_event_time.as_ref() else {
            error_stack::bail!(Error::MissingField {
                field: "max_event_time"
            })
        };
        NaiveDateTime::from_timestamp_opt(timestamp.seconds, timestamp.nanos as u32).ok_or_else(
            || {
                error_stack::report!(Error::InvalidTimestamp {
                    timestamp: timestamp.clone()
                })
            },
        )
    }
}
