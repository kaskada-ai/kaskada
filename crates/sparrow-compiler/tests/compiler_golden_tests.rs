#![warn(
    rust_2018_idioms,
    nonstandard_style,
    future_incompatible,
    clippy::mod_module_files,
    clippy::print_stdout,
    clippy::print_stderr,
    clippy::undocumented_unsafe_blocks
)]

use arrow::datatypes::DataType;
use itertools::Itertools;
use sparrow_api::kaskada::v1alpha::compile_request::ExpressionKind;
use sparrow_api::kaskada::v1alpha::schema::Field;
use sparrow_api::kaskada::v1alpha::Schema;
use sparrow_api::kaskada::v1alpha::{slice_request, SliceRequest};
use sparrow_api::kaskada::v1alpha::{
    CompileRequest, CompileResponse, ComputePlan, ComputeTable, FeatureSet, Formula,
    PerEntityBehavior, TableConfig, TableMetadata,
};
use sparrow_compiler::InternalCompileOptions;
use uuid::Uuid;

#[derive(Debug)]
struct TestScript {
    tables: Vec<TestTable>,
    feature_set: FeatureSet,
}

impl TestScript {
    async fn compile(
        self,
        slice_request: Option<SliceRequest>,
    ) -> error_stack::Result<CompileResponse, sparrow_compiler::Error> {
        let tables = self
            .tables
            .into_iter()
            .map(|table| ComputeTable {
                config: Some(TableConfig::new_with_table_source(
                    &table.name,
                    &table.id,
                    "",
                    Some(""),
                    &table.group_key,
                    &table.grouping,
                )),
                metadata: Some(TableMetadata {
                    schema: Some(table.proto_schema()),
                    file_count: 13,
                }),
                file_sets: vec![],
            })
            .collect();

        sparrow_compiler::compile_proto(
            CompileRequest {
                tables,
                feature_set: Some(self.feature_set),
                slice_request,
                expression_kind: ExpressionKind::Complete as i32,
                experimental: false,
                per_entity_behavior: PerEntityBehavior::All as i32,
            },
            InternalCompileOptions::default(),
        )
        .await
    }
}

#[derive(Debug)]
struct TestTable {
    name: String,
    id: Uuid,
    group_key: String,
    grouping: String,
    fields: Vec<TestField>,
}

impl TestTable {
    fn proto_schema(&self) -> Schema {
        let fields = self
            .fields
            .iter()
            .map(|field| Field {
                name: field.name.clone(),
                data_type: Some((&field.data_type).try_into().unwrap()),
                nullable: false,
            })
            .collect();
        Schema { fields }
    }
}

#[derive(Debug)]
struct TestField {
    name: String,
    data_type: DataType,
}

impl TestField {
    fn new(name: &str, data_type: DataType) -> Self {
        Self {
            name: name.to_owned(),
            data_type,
        }
    }
}

async fn compile(test_script: TestScript, slice_request: Option<SliceRequest>) -> ComputePlan {
    let result = test_script.compile(slice_request).await.unwrap();
    if let Some(plan) = result.plan {
        plan
    } else {
        panic!(
            "Compilation failed with diagnostics:\n{}",
            result
                .fenl_diagnostics
                .unwrap_or_default()
                .fenl_diagnostics
                .into_iter()
                .map(|elt| elt.formatted)
                .format("\n")
        )
    }
}

async fn compile_err(test_script: TestScript) -> String {
    let result = test_script.compile(None).await.unwrap();
    if let Some(plan) = result.plan {
        panic!("Expected no plan with compile errors, but was {plan:?}")
    } else {
        result.fenl_diagnostics.unwrap_or_default().to_string()
    }
}

/// Compile and return the warnings, if any.
async fn compile_warnings(test_script: TestScript) -> String {
    let result = test_script.compile(None).await.unwrap();
    let diagnostics = result.fenl_diagnostics.unwrap_or_default().to_string();

    assert!(
        result.plan.is_some(),
        "Expected plan with only compile warnings:\n\n{diagnostics}",
    );

    diagnostics
}

fn formula(name: &str, formula: &str) -> Formula {
    Formula {
        name: name.to_owned(),
        formula: formula.to_owned(),
        source_location: name.to_owned(),
    }
}

// We can't create a constant for this since it creates `String`
// which isn't (yet) possible.
fn account_sent_table() -> TestTable {
    TestTable {
        name: "Sent".to_owned(),
        id: Uuid::parse_str("111DA01F9ABD4d9d80C702AF85C822A8").unwrap(),
        group_key: "sender".to_owned(),
        grouping: "account".to_owned(),
        fields: vec![
            TestField::new("sender", DataType::UInt64),
            TestField::new("amount", DataType::Float64),
            TestField::new("receiver", DataType::UInt64),
            TestField::new("store", DataType::UInt64),
        ],
    }
}

fn primitive_table() -> TestTable {
    TestTable {
        name: "Primitive".to_owned(),
        id: Uuid::parse_str("222DA01F9ABD4d9d80C702AF85C822A8").unwrap(),
        group_key: "key".to_owned(),
        grouping: "".to_owned(),
        fields: vec![
            TestField::new("key", DataType::UInt64),
            TestField::new("f64", DataType::Float64),
            TestField::new("i64", DataType::Int64),
        ],
    }
}

fn account_received_table() -> TestTable {
    TestTable {
        name: "Received".to_owned(),
        id: Uuid::parse_str("333DA01F9ABD4d9d80C702AF85C822A8").unwrap(),
        group_key: "receiver".to_owned(),
        grouping: "account".to_owned(),
        fields: vec![
            TestField::new("amount", DataType::Float64),
            TestField::new("receiver", DataType::UInt64),
        ],
    }
}

fn store_received_table() -> TestTable {
    TestTable {
        name: "StoreReceived".to_owned(),
        id: Uuid::parse_str("444DA01F9ABD4d9d80C702AF85C822A8").unwrap(),
        group_key: "store_id".to_owned(),
        grouping: "store".to_owned(),
        fields: vec![
            TestField::new("amount", DataType::Float64),
            TestField::new("store_id", DataType::UInt64),
            TestField::new("sender_id", DataType::UInt64),
        ],
    }
}

#[tokio::test]
async fn test_basic_sum() {
    insta::assert_yaml_snapshot!(
        compile(
            TestScript {
                tables: vec![account_sent_table()],
                feature_set: FeatureSet {
                    formulas: vec![
                        formula("f1", "sum(Sent.amount)"),
                        formula("f2", "Sent.amount | sum($input)"),
                    ],
                    query: "{f1, f2}".to_owned(),
                },
            },
            None
        )
        .await
    );
}

#[tokio::test]
async fn test_projection_pushdown() {
    insta::assert_yaml_snapshot!(
        compile(
            TestScript {
                tables: vec![account_sent_table()],
                feature_set: FeatureSet {
                    formulas: vec![],
                    query: "{ amount: Sent | first() | $input.amount  }".to_owned(),
                },
            },
            None
        )
        .await
    );
}

#[tokio::test]
async fn test_unexported() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: vec![formula("f1", "sum(Sent.amount)"), formula("f2", "f1 + 1")],
                query: "{f2}".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_basic_sum_when_event() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: vec![formula("sum_amount", "sum(Sent.amount)")],
                query: "{ sum_amount } | when(Sent.amount > 10)".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}
#[tokio::test]
async fn test_count_since_window() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "count_amount",
                    "count(Sent.amount, window=since(Sent.amount > 10))",
                )],
                query: "{ count_amount }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_basic_sum_when_aggregate() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: vec![formula("sum_amount", "sum(Sent.amount)")],
                query: "{ sum_amount } | when($input.sum_amount > 10)".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_lookup_same_grouping() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), account_received_table()],
            feature_set: FeatureSet {
                formulas: vec![
                    formula("sum_amount", "sum(Sent.amount)"),
                    formula(
                        "sum_received",
                        "lookup(last(Sent.receiver), sum(Received.amount))",
                    ),
                ],
                query: "{ sum_amount, sum_received }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_with_key() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "StoreReceivedBySender",
                    "StoreReceived | with_key($input.sender_id, grouping = \"account\")",
                )],
                query: "{ sum_sent: Sent.amount | sum(), sum_store_received: \
                        StoreReceivedBySender.amount | sum() }"
                    .to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_with_key_default_grouping() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "StoreReceivedBySender",
                    "StoreReceived | with_key($input.sender_id)",
                )],
                query: "{ sum_store_received: StoreReceivedBySender.amount | sum() }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_lookup_same_grouping_with_slicing() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), account_received_table()],
            feature_set: FeatureSet {
                formulas: vec![
                    formula("sum_amount", "sum(Sent.amount)"),
                    formula(
                        "sum_received",
                        "lookup(last(Sent.receiver), sum(Received.amount))",
                    ),
                ],
                query: "{ sum_amount, sum_received }".to_owned(),
            },
        },
        Some(SliceRequest {
            slice: Some(slice_request::Slice::Percent(slice_request::PercentSlice {
                percent: 0.5,
            })),
        }),
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_lookup_diff_grouping() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![
                    formula("sum_amount", "sum(Sent.amount)"),
                    formula(
                        "sum_store_received",
                        "lookup(last(Sent.store), sum(StoreReceived.amount))",
                    ),
                ],
                query: "{ sum_amount, sum_store_received }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_cast_in_record() {
    let compute_plan = compile(
        TestScript {
            tables: vec![primitive_table()],
            feature_set: FeatureSet {
                formulas: Vec::new(),
                query: "{ x: Primitive.i64 } | else ({ x: Primitive.f64 })".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_null_value() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: Vec::new(),
                query: "{ x: null | else(Sent.amount) }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_null_record_field() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: Vec::new(),
                query: "null | else({ a: Sent.amount })".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_null_record() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: Vec::new(),
                query: "{a: Sent.amount, b: null } | else({ a: Sent.amount, b: 7 })".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_valid_time_domains() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "shift",
                    "Sent.amount | shift_until(Sent.amount > 0) | shift_until(Sent.amount < 24)",
                )],
                query: "{ a: shift + shift }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_since_daily() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "daily",
                    "count(Sent.amount, window=since(daily()))",
                )],
                query: "{ a: daily }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_since_daily_multiple_passes() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "count_since",
                    "count(Sent.amount, window=since(daily()))",
                )],
                query: "{ a: count_since | shift_until(Sent.amount > 10) | sum($input, \
                        window=since(daily())) }"
                    .to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_tab_characters_are_allowed() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![],
                query: "	{ a: 	sum(Sent.amount) }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_extend_infers_implicit_input() {
    // Tests that we don't get any diagnostic warnings when implicitly
    // inferring input using an extend.
    //
    // We don't need to assert the plan yaml, since `compile` implicitly
    // asserts that no diagnostic warnings were written.
    let query = "Sent | extend({ sum_extend: sum(Sent.amount) })";
    compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![],
                query: query.to_owned(),
            },
        },
        None,
    )
    .await;
}

#[tokio::test]
async fn test_incompatible_err() {
    let error = compile_err(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "Sent.amount == \"hello\"".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0015]: Incompatible argument types
      --> Query:1:13
      |
    1 | Sent.amount == "hello"
      | ----------- ^^ ------- Type: string
      | |           |
      | |           Incompatible types for call to 'eq'
      | Type: f64
    "###);
}

#[tokio::test]
async fn test_incompatible_shift_functions() {
    let warnings = compile_warnings(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{x: (Sent.amount | shift_until($input > 0)) + (Sent.amount | \
                    shift_until($input > 1)) }"
                .to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(warnings, @r###"
    warning[W2000]: Incompatible time domains
      --> Query:1:45
      |
    1 | {x: (Sent.amount | shift_until($input > 0)) + (Sent.amount | shift_until($input > 1)) }
      |     --------------------------------------- ^ --------------------------------------- Time Domain: ShiftUntil { predicate: 28 }
      |     |                                       |
      |     |                                       Incompatible time domains for operation
      |     Time Domain: ShiftUntil { predicate: 16 }
    "###);
}

#[tokio::test]
async fn test_unused_lhs_pipe() {
    let warnings = compile_warnings(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ x: Sent.amount | sum(Sent.amount) }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(warnings, @r###"
    warning[W2001]: Unused binding
      --> Query:1:18
      |
    1 | { x: Sent.amount | sum(Sent.amount) }
      |                  ^ Left-hand side of pipe not used
    "###);
}

#[tokio::test]
async fn test_unused_let_binding() {
    let warnings = compile_warnings(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "let unused = sum(Sent.amount) in { x: Sent.amount }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(warnings, @r###"
    warning[W2001]: Unused binding
      --> Query:1:5
      |
    1 | let unused = sum(Sent.amount) in { x: Sent.amount }
      |     ^^^^^^ Unused binding 'unused'
    "###);
}

#[tokio::test]
async fn test_incompatible_time_domain_tables_add() {
    let warnings = compile_warnings(TestScript {
        tables: vec![account_sent_table(), account_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{x: Sent.amount + Received.amount }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(warnings, @r###"
    warning[W2000]: Incompatible time domains
      --> Query:1:17
      |
    1 | {x: Sent.amount + Received.amount }
      |     ----------- ^ --------------- Time Domain: Table 'Received'
      |     |           |
      |     |           Incompatible time domains for operation
      |     Time Domain: Table 'Sent'
    "###);
}

#[tokio::test]
async fn test_undefined_field_err() {
    let error = compile_err(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "Sent.amt == 5".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0001]: Illegal field reference
      --> Query:1:6
      |
    1 | Sent.amt == 5
      |      ^^^ No field named 'amt'
      |
      = Nearest fields: 'amount', 'store', 'sender', 'receiver'
    "###);
}

#[tokio::test]
async fn test_invalid_field_ref_base_err() {
    let error = compile_err(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "Sent.amount.foo".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0001]: Illegal field reference
      --> Query:1:1
      |
    1 | Sent.amount.foo
      | ^^^^^^^^^^^ No fields for non-record base type f64
    "###);
}

#[tokio::test]
async fn test_incompatible_grouping_add() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ amount: sum(Received.amount) + sum(StoreReceived.amount) }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[W2001]: Incompatible grouping
      --> Query:1:32
      |
    1 | { amount: sum(Received.amount) + sum(StoreReceived.amount) }
      |           -------------------- ^ ------------------------- Grouping: 'store'
      |           |                    |
      |           |                    Incompatible grouping for operation 'add'
      |           Grouping: 'account'
    "###);
}

#[tokio::test]
async fn test_incompatible_grouping_record() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ sum_account: sum(Received.amount), sum_store: sum(StoreReceived.amount) }"
                .to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[W2001]: Incompatible grouping
      --> Query:1:1
      |
    1 | { sum_account: sum(Received.amount), sum_store: sum(StoreReceived.amount) }
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      | |              |                                |
      | |              |                                Grouping: 'store'
      | |              Grouping: 'account'
      | Incompatible grouping for operation 'record creation'
    "###);
}

#[tokio::test]
async fn test_duplicate_formula() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![
                formula("a", "Received.amount"),
                formula("a", "Received.amount"),
            ],
            query: "{ a }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0004]: Formula already defined
     = Formula name 'a' is defined at index 0 and again at 1.
    "###);
}

#[tokio::test]
async fn test_undefined_formula() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ a }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0006]: Unbound reference
      --> Query:1:3
      |
    1 | { a }
      |   ^ No reference named 'a'
      |
      = Nearest matches: 'Received', 'StoreReceived'
    "###);
}

// Tests that a function that is rewritten into another fenl function
// correctly displays the error on the original signature.
#[tokio::test]
async fn test_invalid_rewritten_function() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ a: sqrt(Sent.amount, Sent.amount) }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0008]: Invalid arguments
      --> Query:1:24
      |
    1 | { a: sqrt(Sent.amount, Sent.amount) }
      |                        ^^^^^^^^^^^ First unexpected argument
      |
      = Expected 1 but got 2
    "###);
}

#[tokio::test]
async fn test_output_type_is_struct_diagnostic() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "Received.amount".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0013]: Invalid output type
     = Output type must be a record, but was f64
    "###);
}
