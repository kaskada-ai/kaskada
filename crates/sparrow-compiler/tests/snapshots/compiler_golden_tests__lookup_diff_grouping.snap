---
source: crates/sparrow-compiler/tests/compiler_golden_tests.rs
expression: compute_plan
---
per_entity_behavior: 1
operations:
  - expressions:
      - arguments: []
        result_type:
          kind:
            Struct:
              fields:
                - name: amount
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
                - name: store_id
                  data_type:
                    kind:
                      Primitive: 10
                  nullable: true
                - name: sender_id
                  data_type:
                    kind:
                      Primitive: 10
                  nullable: true
        output: false
        operator:
          Input:
            producing_operation: 0
            input_column: 0
            interpolation: 1
            column:
              ScanRecord: ~
      - arguments: []
        result_type:
          kind:
            Primitive: 1
        output: false
        operator:
          Literal:
            literal: ~
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: amount
      - arguments:
          - 0
          - 2
        result_type:
          kind:
            Primitive: 13
        output: false
        operator:
          Instruction: field_ref
      - arguments:
          - 3
          - 1
          - 1
        result_type:
          kind:
            Primitive: 13
        output: true
        operator:
          Instruction: sum
    operator:
      Scan:
        table_id:
          high: 4921766025386347933
          low: 9279388510107214504
        schema:
          fields:
            - name: amount
              data_type:
                kind:
                  Primitive: 13
              nullable: true
            - name: store_id
              data_type:
                kind:
                  Primitive: 10
              nullable: true
            - name: sender_id
              data_type:
                kind:
                  Primitive: 10
              nullable: true
        slice_plan:
          table_name: StoreReceived
          slice: ~
  - expressions:
      - arguments: []
        result_type:
          kind:
            Struct:
              fields:
                - name: sender
                  data_type:
                    kind:
                      Primitive: 10
                  nullable: true
                - name: amount
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
                - name: receiver
                  data_type:
                    kind:
                      Primitive: 10
                  nullable: true
                - name: store
                  data_type:
                    kind:
                      Primitive: 10
                  nullable: true
        output: false
        operator:
          Input:
            producing_operation: 1
            input_column: 0
            interpolation: 1
            column:
              ScanRecord: ~
      - arguments: []
        result_type:
          kind:
            Primitive: 1
        output: false
        operator:
          Literal:
            literal: ~
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: store
      - arguments:
          - 0
          - 2
        result_type:
          kind:
            Primitive: 10
        output: false
        operator:
          Instruction: field_ref
      - arguments:
          - 3
          - 1
          - 1
        result_type:
          kind:
            Primitive: 10
        output: true
        operator:
          Instruction: last
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: amount
      - arguments:
          - 0
          - 5
        result_type:
          kind:
            Primitive: 13
        output: false
        operator:
          Instruction: field_ref
      - arguments:
          - 6
          - 1
          - 1
        result_type:
          kind:
            Primitive: 13
        output: true
        operator:
          Instruction: sum
    operator:
      Scan:
        table_id:
          high: 1233317930569911709
          low: 9279388510107214504
        schema:
          fields:
            - name: sender
              data_type:
                kind:
                  Primitive: 10
              nullable: true
            - name: amount
              data_type:
                kind:
                  Primitive: 13
              nullable: true
            - name: receiver
              data_type:
                kind:
                  Primitive: 10
              nullable: true
            - name: store
              data_type:
                kind:
                  Primitive: 10
              nullable: true
        slice_plan:
          table_name: Sent
          slice: ~
  - expressions:
      - arguments: []
        result_type:
          kind:
            List:
              name: item
              item_type:
                kind:
                  Primitive: 10
              nullable: true
        output: true
        operator:
          Input:
            producing_operation: 1
            input_column: 2
            interpolation: 1
            column:
              KeyColumn: 3
    operator:
      LookupRequest:
        primary_operation: 1
        foreign_key_hash:
          producing_operation: 1
          input_column: 3
          interpolation: 2
          column:
            ProducerExpression: 4
  - expressions:
      - arguments: []
        result_type:
          kind:
            List:
              name: item
              item_type:
                kind:
                  Primitive: 10
              nullable: true
        output: true
        operator:
          Input:
            producing_operation: 2
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 0
      - arguments: []
        result_type:
          kind:
            Primitive: 13
        output: true
        operator:
          Input:
            producing_operation: 0
            input_column: 3
            interpolation: 2
            column:
              ProducerExpression: 4
    operator:
      Merge:
        left: 2
        right: 0
  - expressions:
      - arguments: []
        result_type:
          kind:
            Primitive: 13
        output: true
        operator:
          Input:
            producing_operation: 3
            input_column: 4
            interpolation: 2
            column:
              ProducerExpression: 1
    operator:
      LookupResponse:
        foreign_operation: 3
        requesting_key_hash:
          producing_operation: 3
          input_column: 3
          interpolation: 1
          column:
            ProducerExpression: 0
  - expressions:
      - arguments: []
        result_type:
          kind:
            Primitive: 24
        output: false
        operator:
          LateBound: 1
      - arguments: []
        result_type:
          kind:
            Primitive: 13
        output: false
        operator:
          Input:
            producing_operation: 4
            input_column: 3
            interpolation: 2
            column:
              ProducerExpression: 0
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: sum_store_received
      - arguments: []
        result_type:
          kind:
            Primitive: 13
        output: false
        operator:
          Input:
            producing_operation: 1
            input_column: 4
            interpolation: 2
            column:
              ProducerExpression: 7
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: sum_amount
      - arguments:
          - 4
          - 3
          - 2
          - 1
        result_type:
          kind:
            Struct:
              fields:
                - name: sum_amount
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
                - name: sum_store_received
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
        output: true
        operator:
          Instruction: record
      - arguments:
          - 5
        result_type:
          kind:
            Primitive: 24
        output: false
        operator:
          Instruction: time_of
      - arguments:
          - 6
          - 0
        result_type:
          kind:
            Primitive: 2
        output: true
        operator:
          Instruction: gte
    operator:
      Merge:
        left: 1
        right: 4
  - expressions:
      - arguments: []
        result_type:
          kind:
            Struct:
              fields:
                - name: sum_amount
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
                - name: sum_store_received
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
        output: true
        operator:
          Input:
            producing_operation: 5
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 5
    operator:
      Select:
        input: 5
        condition:
          producing_operation: 5
          input_column: 4
          interpolation: 1
          column:
            ProducerExpression: 7
primary_grouping: account
primary_grouping_key_type:
  kind:
    Primitive: 10

